File: lib/core/normalize.dart
/// Metin normalizasyonu. Hepsi saf fonksiyondur: Firebase, widget ya da
/// başka bir bağımlılığı yoktur, doğrudan birim testle kaplanır.
library;

final RegExp _whitespace = RegExp(r'\s+');

/// Realtime Database anahtarlarında kullanılamayan karakterler.
final RegExp _invalidKeyChars = RegExp(r'[.$#\[\]/\x00-\x1F\x7F]');

/// Türkçe kurallarına göre büyük harfe çevirir.
///
/// Dart'ın `toUpperCase()` metodu locale tanımaz ve `i` harfini `I` yapar;
/// Türkçe'de karşılığı `İ` olmalıdır. `ı` harfi de `I` olmalıdır.
String turkishUpperCase(String value) {
  final StringBuffer buffer = StringBuffer();
  for (final int rune in value.runes) {
    if (rune == 0x69) {
      buffer.writeCharCode(0x130); // i -> İ
    } else if (rune == 0x131) {
      buffer.writeCharCode(0x49); // ı -> I
    } else {
      buffer.write(String.fromCharCode(rune).toUpperCase());
    }
  }
  return buffer.toString();
}

/// Plakayı büyük harfe çevirir ve fazla boşlukları temizler.
///
/// `"  34 abc  123 "` -> `"34 ABC 123"`
String normalizePlate(String raw) {
  return turkishUpperCase(raw.trim()).replaceAll(_whitespace, ' ');
}

/// Sürücü adındaki baş/son ve tekrarlı boşlukları temizler.
/// Büyük/küçük harfe dokunmaz.
String normalizeDriverName(String raw) {
  return raw.trim().replaceAll(_whitespace, ' ');
}

/// Grup adını normalleştirir.
///
/// Grup adı aynı zamanda veritabanı anahtarıdır; anahtarlarda `.`, `$`, `#`,
/// `[`, `]`, `/` ve kontrol karakterleri kullanılamaz, bu yüzden atılırlar.
String normalizeGroupName(String raw) {
  return raw
      .trim()
      .replaceAll(_whitespace, ' ')
      .replaceAll(_invalidKeyChars, '')
      .trim();
}

/// Yöneticinin yazdığı "görünür gruplar" metnini listeye çevirir.
///
/// Virgülle ayrılır, boşlar atılır, yinelenenler tekilleştirilir, sıra korunur.
/// `"Merkez, Kaman ,, Merkez"` -> `["Merkez", "Kaman"]`
List<String> parseGroupList(String raw) {
  final List<String> result = <String>[];
  for (final String part in raw.split(',')) {
    final String name = normalizeGroupName(part);
    if (name.isEmpty) continue;
    if (result.contains(name)) continue;
    result.add(name);
  }
  return result;
}

/// Listeyi düzenleme kutusunda gösterilecek metne çevirir.
String formatGroupList(Iterable<String> groups) => groups.join(', ');

File: panel/js/normalize.js
// Metin normalizasyonu. lib/core/normalize.dart dosyasının karşılığı.
// Hepsi saf fonksiyondur.

const WHITESPACE = /\s+/g;

// Realtime Database anahtarlarında kullanılamayan karakterler.
const INVALID_KEY_CHARS = /[.$#[\]/\x00-\x1F\x7F]/g;

// Türkçe kurallarına göre büyük harfe çevirir.
//
// JS'in toUpperCase() metodu da Dart'ınki gibi locale tanımaz: `i` harfini `I`
// yapar, Türkçe'de karşılığı `İ` olmalıdır. `ı` harfi de `I` olmalıdır.
export function turkishUpperCase(value) {
  let result = '';
  for (const char of value) {
    if (char === 'i') result += 'İ';
    else if (char === 'ı') result += 'I';
    else result += char.toUpperCase();
  }
  return result;
}

// Baş/son ve tekrarlı boşlukları temizler, büyük/küçük harfe dokunmaz.
export function normalizeDriverName(raw) {
  return raw.trim().replace(WHITESPACE, ' ');
}

// Grup adı aynı zamanda veritabanı anahtarıdır; yasak karakterler atılır.
export function normalizeGroupName(raw) {
  return raw.trim().replace(WHITESPACE, ' ').replace(INVALID_KEY_CHARS, '').trim();
}

// E-posta biçimi. Formun kendi kontrolüdür, güvenlik sınırı değildir;
// asıl doğrulamayı Firebase Auth yapar.
export function isValidEmail(raw) {
  return /^[^@\s]+@[^@\s]+\.[^@\s]+$/.test(raw.trim());
}
